crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,359 @@
+        "https://api.github.com/repos/{}/{}/releases/tags/v{}",
+        TYPST_GITHUB_ORG, TYPST_REPO, version
+    );
+
+    download_release(&url)
+}
+
+/// Specifies the url for the latest release.
+fn latest_release() -> StrResult<Release> {
+    let url = format!(
+        "https://api.github.com/repos/{}/{}/releases/latest",
+        TYPST_GITHUB_ORG, TYPST_REPO
+    );
+
+    download_release(&url)
+}
+
+/// Downloads and parses a GitHub release from the Typst repository.
+fn download_release(url: &str) -> StrResult<Release> {
+    match ureq::get(url).call() {
+        Ok(response) => {
+            Ok(response.into_json().map_err(|_| "unable to get json from response")?)
+        }
+        Err(ureq::Error::Status(404, _)) => {
+            // TODO: maybe make enum error type in typst::diag for this?
+            bail!("release not found")
+        }
+        // TODO: same as above
+        Err(_) => bail!("network failed"),
+    }
+}
+
+/// Sorts through the assets from a given `Release` and picks the right one
+/// for this target platform.
+///
+/// Returns a compressed archive that contains the Typst pre-compiled binary.
+fn download_asset_archive(release: &Release) -> StrResult<Archive> {
+    let asset_needed = asset_needed()?;
+
+    let asset = release
+        .assets
+        .iter()
+        .find(|a| a.name.starts_with(asset_needed))
+        .ok_or("could not find release for your target platform.")?;
+
+    let response = match ureq::get(&asset.browser_download_url).call() {
+        Ok(response) => response,
+        Err(ureq::Error::Status(404, _)) => {
+            // TODO: maybe make enum error type in typst::diag for this?
+            bail!("asset not found");
+        }
+        // TODO: same as above
+        Err(_) => bail!("network failed"),
+    };
+
+    let len = response.header("Content-Length").unwrap().parse().unwrap();
+    let mut buffer = Vec::with_capacity(len);
+    response.into_reader().read_to_end(&mut buffer).unwrap();
+
+    let extension = Path::new(&asset.name)
+        .extension()
+        .and_then(OsStr::to_str)
+        .ok_or("failed to get extension from archive")?
+        .into();
+
+    Ok(Archive {
+        extension,
+        buffer: BufReader::new(Cursor::new(buffer)),
+    })
+}
+
+/// Unpacks the asset archive in-memory and writes the uncompressed contents
+/// into a buffer.
+///
+/// In-memory refers to that the physical archive is never written to disk, only
+/// the Typst CLI binary will be plucked from the archive and written to disk at
+/// a later stage, the rest of the archive is discarded.
+fn unpack_archive(archive: Archive) -> StrResult<Vec<u8>> {
+    match archive.extension {
+        Extension::Zip => {
+            let mut zip = zip::ZipArchive::new(archive.buffer)
+                .map_err(|err| eco_format!("Error opening zip archive: {}", err))?;
+            let mut binary = zip
+                .by_name(&format!("{}/typst.exe", asset_needed()?))
```

Comment:
Isn't this only going to work for Windows targets (due to the extension)? Some branching might be needed here

Reply:
I figured since only the `typst-x86_64-pc-windows-msvc` is zipped that it was _somewhat_ safe to assume it would contain an `.exe`. My assumption is not that safe but considering any chance to the release naming convention would break the installer.. I'm not sure how far I should go in future proofing but I'm looking into a good solution.